x-pack/platform/plugins/shared/inference/server/chat_complete/anonymization/deanonymize_message.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  Message,
+  ChatCompletionEvent,
+  MessageRole,
+  AnonymizationEntity,
+  AnonymizationOutput,
+} from '@kbn/inference-common';
+import {
+  ChatCompletionEventType,
+  ChatCompletionChunkEvent,
+} from '@kbn/inference-common/src/chat_complete/events';
+import { OperatorFunction, mergeMap, filter, of, identity } from 'rxjs';
+import { deanonymize } from './deanonymize';
+
+export function deanonymizeMessage(
+  anonymization: AnonymizationOutput
+): OperatorFunction<ChatCompletionEvent, ChatCompletionEvent> {
+  if (!anonymization.anonymizations.length) {
+    return identity;
+  }
+  const anonymizationEntityLookup: Record<string, AnonymizationEntity> = {};
+
+  anonymization.anonymizations.forEach((item) => {
+    anonymizationEntityLookup[item.entity.mask] = item.entity;
+  });
+
+  return (source$) => {
+    // Buffer to store chunks
+    const chunks: ChatCompletionChunkEvent[] = [];
+
+    return source$.pipe(
+      // Filter out regular chunks without emitting them
+      filter((event) => {
+        if (event.type === ChatCompletionEventType.ChatCompletionChunk) {
+          chunks.push(event as ChatCompletionChunkEvent);
```

Comment:
I wouldn't commit side-effects here, or make it very explicit that you're using side-effects by using new Observable(...). Do you even need to keep track of `chunks`?